stumpy/ostinato.py

```diff
@@ -0,0 +1,272 @@
+
+    ss_ind : int
+        Start index of subsequence in `ts_ind` that has radius `rad`
+
+    m : int
+        Window size
+
+    Returns
+    -------
+    rad : float
+        Radius of the most central consensus motif
+
+    ts_ind : int
+        Index of time series which contains the most central consensus motif
+
+    ss_ind : int
+        Start index of most central consensus motif within the time series `ts_ind`
+        that contains it
+
+    Notes
+    -----
+    <https://www.cs.ucr.edu/~eamonn/consensus_Motif_ICDM_Long_version.pdf>
+
+    See Table 2
+
+    The ostinato algorithm proposed in the paper finds the best radius
+    in `tss`. Intuitively, the radius is the minimum distance of a
+    subsequence to encompass at least one nearest neighbor subsequence
+    from all other time series. The best radius in `tss` is the minimum
+    radius amongst all radii. Some data sets might contain multiple
+    subsequences which have the same optimal radius.
+    The greedy Ostinato algorithm only finds one of them, which might
+    not be the most central motif. The most central motif amongst the
+    subsequences with the best radius is the one with the smallest mean
+    distance to nearest neighbors in all other time series. To find this
+    central motif it is necessary to search the subsequences with the
+    best radius via `stumpy.ostinato._get_central_motif`
+    """
+    k = len(tss)
+
+    # Ostinato hit: get nearest neighbors' distances and indices
+    q_ost = tss[ts_ind][ss_ind : ss_ind + m]
+    ss_ind_nn_ost, d_ost = _across_series_nearest_neighbors(q_ost, tss)
+
+    # Alternative candidates: Distance to ostinato hit equals best radius
+    ts_ind_alt = np.flatnonzero(np.isclose(d_ost, rad))
+    ss_ind_alt = ss_ind_nn_ost[ts_ind_alt]
+    d_alt = np.zeros((len(ts_ind_alt), k), dtype=float)
+    for i, (tsi, ssi) in enumerate(zip(ts_ind_alt, ss_ind_alt)):
+        q = tss[tsi][ssi : ssi + m]
+        _, d_alt[i] = _across_series_nearest_neighbors(q, tss)
+    rad_alt = np.max(d_alt, axis=1)
+    d_mean_alt = np.mean(d_alt, axis=1)
+
+    # Alternatives with same radius and lower mean distance
+    alt_better = np.logical_and(
+        np.isclose(rad_alt, rad),
+        d_mean_alt < d_ost.mean(),
+    )
+    # Alternatives with same radius and same mean distance
+    alt_same = np.logical_and(
+        np.isclose(rad_alt, rad),
+        np.isclose(d_mean_alt, d_ost.mean()),
+    )
+    if np.any(alt_better):
+        ts_ind_alt = ts_ind_alt[alt_better]
+        ss_ind_alt = ss_ind_alt[alt_better]
+        d_mean_alt = d_mean_alt[alt_better]
+        i_alt_best = np.argmin(d_mean_alt)
+        return rad, ts_ind_alt[i_alt_best], ss_ind_alt[i_alt_best]
+    elif np.any(alt_same):
+        # Default to the first match in the list of time series
+        ts_ind_alt = ts_ind_alt[alt_same]
+        ss_ind_alt = ss_ind_alt[alt_same]
+        i_alt_first = np.argmin(ts_ind_alt)
+        if ts_ind_alt[i_alt_first] < ts_ind:
```

Comment:
This amount of nested branching logic concerns me (from a testing standpoint). I can try to spend some time in the future to see if it can be simplified but just leaving a note for my future self here.

Reply:
See #285

Reply:
Oh, in case it matters, I would not put `naive_get_central_motif` in `naive.py` because `naive_get_central_motif` is only used by `test_ostinato.py` and no other test file. Functions in `naive.py` are strictly for naive implementations that are used across multiple different test files.

Reply:
Yeah, when I wrote this I also got that code smell. I'll see if i can resolve it ;)